crates/provider/src/layers/cache.rs

```diff
@@ -0,0 +1,436 @@
+}
+
+#[cfg(test)]
+mod tests {
+    use std::str::FromStr;
+
+    use crate::ProviderBuilder;
+    use alloy_network::TransactionBuilder;
+    use alloy_node_bindings::Anvil;
+    use alloy_primitives::{Bytes, FixedBytes};
+    use alloy_rpc_types_eth::{BlockId, TransactionRequest};
+
+    use super::*;
+
+    #[tokio::test]
+    async fn test_cache_provider() {
+        let cache = CacheLayer::new(100);
+        let anvil = Anvil::new().block_time_f64(0.3).spawn();
+        let provider = ProviderBuilder::default().layer(cache).on_http(anvil.endpoint_url());
+
+        let path = PathBuf::from_str("./rpc-cache-block-by-number.txt").unwrap();
+        provider.load_cache(path.clone()).unwrap();
+
+        let blk = provider.get_block_by_number(0.into(), true).await.unwrap();
+        let blk2 = provider.get_block_by_number(0.into(), true).await.unwrap();
+        assert_eq!(blk, blk2);
+
+        tokio::time::sleep(tokio::time::Duration::from_millis(500)).await;
+
+        let latest_block_num = provider.get_block_number().await.unwrap();
+        let blk3 = provider.get_block_by_number(latest_block_num.into(), true).await.unwrap();
+        let blk4 = provider.get_block_by_number(latest_block_num.into(), true).await.unwrap();
+        assert_eq!(blk3, blk4);
+
+        provider.save_cache(path).unwrap();
+    }
+
+    #[tokio::test]
+    async fn test_get_block() {
+        let cache = CacheLayer::new(100);
+        let anvil = Anvil::new().block_time_f64(0.3).spawn();
+        let provider = ProviderBuilder::default().layer(cache).on_http(anvil.endpoint_url());
+
+        let path = PathBuf::from_str("./rpc-cache-block-by-hash.txt").unwrap();
+        provider.load_cache(path.clone()).unwrap();
+
+        let block = provider.get_block(0.into(), BlockTransactionsKind::Full).await.unwrap(); // Received from RPC.
+        let block2 = provider.get_block(0.into(), BlockTransactionsKind::Full).await.unwrap(); // Received from cache.
+        assert_eq!(block, block2);
+
+        tokio::time::sleep(tokio::time::Duration::from_millis(500)).await;
+
+        let latest_block =
+            provider.get_block(BlockId::latest(), BlockTransactionsKind::Full).await.unwrap(); // Received from RPC.
+        let latest_hash = latest_block.unwrap().header.hash;
+
+        let block3 =
+            provider.get_block_by_hash(latest_hash, BlockTransactionsKind::Full).await.unwrap(); // Received from RPC.
+        let block4 =
+            provider.get_block_by_hash(latest_hash, BlockTransactionsKind::Full).await.unwrap(); // Received from cache.
+        assert_eq!(block3, block4);
+
+        provider.save_cache(path).unwrap();
+    }
+
+    #[tokio::test]
+    async fn test_get_proof() {
+        let cache = CacheLayer::new(100);
+        let anvil = Anvil::new().block_time_f64(0.3).spawn();
+        let provider = ProviderBuilder::default()
+            // .with_recommended_fillers() - TODO: Issue here. Layer doesn't work with fillers. Fix.
```

Comment:
Enabling fillers would wrap the `CacheProvider` or any other layer provider in a `FillProvider`. Making methods on `CacheProvider` inaccessible. i.e `provider.load_cache(path)` won't work but `provider.inner.load_cache(path)` works.

This is not ideal.